### ISSUE TITLE
Add type annotations?

### ISSUE THREAD

**Issue by astrojuanlu** [poliastro/poliastro#1230]
Comes from #1223:

> Although I have to admit I am a big fan of type annotations to empower mypy to detect as much problems as possible.

Yes! I'm a fan as well, but I have concerns about the verbosity of the syntax, the fact that MyPy has some limitations when working with stubs that are either very difficult to fix or not a priority for the developers (https://github.com/python/mypy/issues/5028), the changes to come in Python 3.10, and so forth. Python type annotations and gradual typing have evolved in a very organic way: first annotations were introduced without any particular intent, then they were made for gradual typing, then they just simplified the syntax... Even though I am normally happy with living in the bleeding edge in terms of libraries and such, for type annotations I have some reservations. That's why I didn't jump immediately and said "let's merge this".

Besides, there's the question of annotating _all_ the library to be consistent, and finally we have another problem which is the interaction with `astropy.units`.

As a result, before starting the path of annotating poliastro, I would like to see some discussion, and a way forward. For example: should we start by slowly making MyPy more strict, and adding types more gradually? This is the "full MyPy strict mode" I've been using in other projects:

```
[mypy]
files =
    src/**/*.py,
    tests/**/*.py
warn_redundant_casts = True
warn_unused_configs = True
pretty = True
show_error_codes = True

disallow_any_generics = True
disallow_subclassing_any = True
disallow_untyped_calls = True
disallow_incomplete_defs = True
check_untyped_defs = True
disallow_untyped_decorators = True
no_implicit_optional = True
warn_unused_ignores = True
warn_return_any = True
no_implicit_reexport = True

# More strict checks for library code
[mypy-poliastro]
disallow_untyped_defs = True
```

which is very far from what we have now:

https://github.com/poliastro/poliastro/blob/5ee92de8b620efbf215d3fca1ba248beda41428f/mypy.ini#L1-L4

I think we could start by enabling some of these configs (the easiest ones), doing small tweaks, and then seeing where is the "inflection point" where we need to annotate large chunks of the code.

And one last note: did the recent changes in Astropy made the interaction between MyPy and `quantity_input` easier, or more robust, or more appealing? Do we need to wait until Astropy 4.3 for any particular functionality?

**Comment by ismaelJimenez** [comment]
Thanks @astrojuanlu for opening this issue. The verbosity of type annotations, as you said, is definitely an issue although overall I have found adding type annotations quite useful.

In case you would like to go for adding type annotations, one proposal would be to follow a guideline similar to Pandas (https://pandas.pydata.org/pandas-docs/stable/development/contributing.html) were the new code is required to have type annotations, so they are adding type annotations incrementally instead of going for a big refactoring.

With respect to the quantity_input from astropy, I would recommend to use it whenever the method inputs are unit types. For those methods, if there are parameters that are not unit types we could also try to annotate them although I have found that sometimes quantity_input has conflicts with type annotations. My proposal is that in case of conflict the use of quantity_input trumps general type annotations.

I also think that it would be a good idea to go step by step towards a mypy configuration like the one you proposed.

What do you think?

**Comment by s-m-e** [comment]
A little side-note: If you add type annotations, you may also consider using [typeguard](https://typeguard.readthedocs.io/en/latest/) for run-time type checking. It's slow like hell, but it is absolutely useful for testing. For production, you can always deactivate it by running Python in [optimized mode](https://typeguard.readthedocs.io/en/latest/userguide.html#using-type-checker-functions). Run-time type checking actually goes a few steps further than what static code analysis can provide and show you even more bugs (than you ever wanted to find).

**Comment by astrojuanlu** [comment]
I have yet one more gripe with annotations, and is that they are still in flux. In fact, they had to postpone a change from Python 3.10 to Python 3.11 because otherwise lots of web stuff would break https://mail.python.org/archives/list/python-dev@python.org/message/CLVXXPQ2T2LQ5MP2Y53VVQFCXYWQJHKZ/ . In fact, our own `czml3` will be broken by this.

Therefore, with all the caveats above, and taking a (perhaps overly conservative) stance, my vote for type annotations is "not yet".

**Comment by s-m-e** [comment]
Speaking as someone who even runs Linux LTS distributions beyond their due date, I am always in favor of a very stable, conservative approach ;) Or I am simply lacking the time to upgrade ... Besides, I agree: This stuff being in flux is annoying and makes it unreliable. Just the fact that you had to import a special alias for lists, i.e. `List`, so you could write `List[int]` and now you're supposed to write `list[int]` all the sudden (in 3.9 or 3.10?), which is actually confusing in multiple ways, drives me nuts.

Having said that, the question is what you want to use type annotations for and how much you want to rely on them. Here is where the pros come in: Amazing support in IDEs, i.e. way faster development and fewer bugs. Besides, both static type checking and run time checks actually provide an amazing amount of insight into your code while *testing* and *debugging*. But it's like with coverage: Is it worth it to have 100% coverage at all cost? Or is it merely a tool to be on the safe side for most of your critical code? Besides, how often do you run coverage analysis in production setups?

I'd say do not rely on every last detail of the spec and use it with care and in the right quantity. Then it actually makes sense (although it is still in flux). I have gone through this process and ended up trying it in a [project of mine](https://github.com/pleiszenburg/bewegung). It is worth it.

**Comment by s-m-e** [comment]
Have I mentioned that Sphinx automatically picks up the notations and uses them when generating API docs? Less complex/confusing doc strings, actually more readable code, ...

**Comment by astrojuanlu** [comment]
TIL: https://pypi.org/project/typing-copilot/
